File: modules/modules/audio-interruption/android/src/main/java/expo/modules/audiointerruption/AudioInterruptionModule.kt
package expo.modules.audiointerruption

import android.content.Context
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class AudioInterruptionModule : Module() {
  private var audioManager: AudioManager? = null
  private var focusListener: OnAudioFocusChangeListener? = null

  override fun definition() = ModuleDefinition {
    Name("AudioInterruption")

    Events("onInterruptionBegan", "onInterruptionEnded")

    OnCreate {
      audioManager = appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as AudioManager

      focusListener = OnAudioFocusChangeListener { focusChange ->
        when (focusChange) {
          AudioManager.AUDIOFOCUS_LOSS,
          AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> sendEvent("onInterruptionBegan", mapOf<String, Any>())
          AudioManager.AUDIOFOCUS_GAIN -> sendEvent("onInterruptionEnded", mapOf<String, Any>())
        }
      }
    }

    OnDestroy {
      focusListener = null
    }
  }
}

File: modules/audio-interruption/src/AudioInterruptionModule.ts
import { requireNativeModule } from "expo-modules-core";

const AudioInterruptionModule = requireNativeModule("AudioInterruption");

export function addInterruptionBeganListener(listener: () => void) {
  return AudioInterruptionModule.addListener("onInterruptionBegan", listener);
}

export function addInterruptionEndedListener(listener: () => void) {
  return AudioInterruptionModule.addListener("onInterruptionEnded", listener);
}
